src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -1023,3 +1048,497 @@ address StubGenerator::generate_poly1305_processBlocks() {
+  __ vpxor(P0H, P0H, P0H, Assembler::AVX_256bit);
+  __ vpxor(P1L, P1L, P1L, Assembler::AVX_256bit);
+  __ vpxor(P1H, P1H, P1H, Assembler::AVX_256bit);
+  __ vpxor(P2L, P2L, P2L, Assembler::AVX_256bit);
+  __ vpxor(P2H, P2H, P2H, Assembler::AVX_256bit);
+
+  // Calculate partial products
+  // p0 = a2xr1'
+  // p1 = a2xr2'
+  // p2 = a2xr0
+  __ vpmadd52luq(P0L, A2, R1P, Assembler::AVX_256bit);
+  __ vpmadd52huq(P0H, A2, R1P, Assembler::AVX_256bit);
+
+  __ vpmadd52luq(P1L, A2, R2P, Assembler::AVX_256bit);
+  __ vpmadd52huq(P1H, A2, R2P, Assembler::AVX_256bit);
+
+  __ vpmadd52luq(P0L, A0, R0, Assembler::AVX_256bit);
+  __ vpmadd52huq(P0H, A0, R0, Assembler::AVX_256bit);
+
+  // p2 = a2xr0
+  __ vpmadd52luq(P2L, A2, R0, Assembler::AVX_256bit);
+  __ vpmadd52huq(P2H, A2, R0, Assembler::AVX_256bit);
+
+  __ vpmadd52luq(P1L, A0, R1, Assembler::AVX_256bit);
+  __ vpmadd52huq(P1H, A0, R1, Assembler::AVX_256bit);
+
+  __ vpmadd52luq(P0L, A1, R2P, Assembler::AVX_256bit);
+  __ vpmadd52huq(P0H, A1, R2P, Assembler::AVX_256bit);
+
+  __ vpmadd52luq(P2L, A0, R2, Assembler::AVX_256bit);
+  __ vpmadd52huq(P2H, A0, R2, Assembler::AVX_256bit);
+
+  // Carry propgation (first pass)
+  __ vpsrlq(YTMP1, P0L, 44, Assembler::AVX_256bit);
+  __ vpsllq(P0H, P0H, 8, Assembler::AVX_256bit);
+  //
+  __ vpmadd52luq(P1L, A1, R0, Assembler::AVX_256bit);
+  __ vpmadd52huq(P1H, A1, R0, Assembler::AVX_256bit);
+  // Carry propagation (first pass) - continue
+  __ vpand(A0, P0L, ExternalAddress(poly1305_mask44()), Assembler::AVX_256bit, rscratch); // Clear top 20 bits
+  __ vpaddq(P0H, P0H, YTMP1, Assembler::AVX_256bit);
+  //
+  __ vpmadd52luq(P2L, A1, R1, Assembler::AVX_256bit);
+  __ vpmadd52huq(P2H, A1, R1, Assembler::AVX_256bit);
+
+  // Carry propagation (first pass) - continue
+  __ vpaddq(P1L, P1L, P0H, Assembler::AVX_256bit);
+  __ vpsllq(P1H, P1H, 8, Assembler::AVX_256bit);
+  __ vpsrlq(YTMP1, P1L, 44, Assembler::AVX_256bit);
+  __ vpand(A1, P1L, ExternalAddress(poly1305_mask44()), Assembler::AVX_256bit, rscratch); // Clear top 20 bits
+
+  __ vpaddq(P2L, P2L, P1H, Assembler::AVX_256bit);
+  __ vpaddq(P2L, P2L, YTMP1, Assembler::AVX_256bit);
+  __ vpand(A2, P2L, ExternalAddress(poly1305_mask42()), Assembler::AVX_256bit, rscratch); // Clear top 22 bits
+  __ vpsrlq(YTMP1, P2L, 42, Assembler::AVX_256bit);
+  __ vpsllq(P2H, P2H, 10, Assembler::AVX_256bit);
+  __ vpaddq(P2H, P2H, YTMP1, Assembler::AVX_256bit);
+
+  // Carry propagation (second pass)
+  // Multiply by 5 the highest bits (above 130 bits)
+  __ vpaddq(A0, A0, P2H, Assembler::AVX_256bit);
+  __ vpsllq(P2H, P2H, 2, Assembler::AVX_256bit);
+  __ vpaddq(A0, A0, P2H, Assembler::AVX_256bit);
+
+  __ vpsrlq(YTMP1, A0, 44, Assembler::AVX_256bit);
+  __ vpand(A0, A0, ExternalAddress(poly1305_mask44()), Assembler::AVX_256bit, rscratch); // Clear top 20 bits
+  __ vpaddq(A1, A1, YTMP1, Assembler::AVX_256bit);
+
+}
+
+void StubGenerator::poly1305_msg_mul_reduce_vec4_avx2(
```

Comment:
This function really should not exist. It is just a duplicate of `poly1305_mul_reduce_vec4_avx2` and `poly1305_limbs_avx2` interleaved for 'maximum confusion'.. 

(The easiest 'trick' to detangling I found.. Most of the interleaved instructions are not dependent, so you can move them around. 
So just move 2-3 lines, run fuzz.. move 2-3 lines, run fuzz.. eventually you have correct code that looks like:
```
poly1305_limbs_avx2()
poly1305_multiply4_avx2()
```
And if you made a mistake, you only have to move the 2-3 lines back; hardest part of getting `poly1305_multiply4_avx2` correct is done, refactoring using this process is fairly mindless and error-free)

Reply:
Sticking to Pablo's implementation for ease of debugging and saving time